### PR TITLE
[red-knot] Remove `Type::None`

### LINKED ISSUE #13670: [red-knot] remove `Type::None` in favor of always using `Type::Instance(_typeshed_.NoneType)`

### PR DESCRIPTION
**Pull request by sharkdp**
## Summary

Removes `Type::None` in favor of `KnownClass::NoneType.to_instance(…)`.

closes #13670

## Test Plan

Existing tests pass.

### CODE REVIEW

File: crates/red_knot_python_semantic/src/types.rs
Comment (AlexWaygood) [inline crates/red_knot_python_semantic/src/types.rs:530]: We should be able to remove this branch too once we understand that `NoneType` is a subclass of `object` (my MRO PR is coming soon!!)

File: crates/red_knot_python_semantic/src/types/infer.rs
Comment (sharkdp) [reply crates/red_knot_python_semantic/src/types/infer.rs:3637]: Oh. Forgot that I had planned ahead!

File: crates/red_knot_python_semantic/src/types.rs
Comment (sharkdp) [reply crates/red_knot_python_semantic/src/types.rs:530]: Note that this is about `NoneType <: NoneType`. The `X <: object` case is handled further above.

I'm actually surprised that `NoneType <: NoneType` doesn't work without this. We have a `other == self` check in `is_subclass_of` on the `ClassType`s. Is that not working correctly? Is it actually checking for equal classes or for some other identity (salsa).

> my MRO PR is coming soon!!

That will definitely help us remove some cases!

Comment (sharkdp) [reply crates/red_knot_python_semantic/src/types.rs]: I pushed a commit that avoids this branch and handles it inside `is_sublass_of`, by also comparing on known classes. I think that `self == other` check there is a little misleading.

Comment (carljm) [reply crates/red_knot_python_semantic/src/types.rs:530]: `self == other` is indeed just checking salsa IDs. But Salsa IDs are assigned by Salsa interning, which interns based on struct equality, so that should be the same thing as an `Eq` check.

It would be good to understand what's happening here. Let me know if you'd like another pair of eyes on it. Maybe it's related to the fact that there are multiple `NoneType` in typeshed? (The one in `_typeshed` and the one in `types` for more recent Python versions.)

Also, I think this branch should go into `is_equivalent_to` instead of `is_subtype_of`.

**Review by carljm — APPROVED** [review]
Looks great!

I wonder how much of the perf regression is due to the fact that currently I think every `None` ends up being a union of the `None` defined in `_typeshed` and the one defined in`types`? This could mean we spend a lot more time in recursive union handling for things like `is_subtype_of`? If so, this should be largely mitigated by `sys.version_info` checking.

**Comment by MichaReiser** [comment]
I took a look at the perf regression ([main](https://profiler.firefox.com/from-url/http%3A%2F%2F127.0.0.1%3A3000%2F28rfz7p9bj3skawjfq52bbwqyjaf96mlwny5pcm%2Fprofile.json/calltree/?globalTrackOrder=0w9&hiddenGlobalTracks=0w35w9&hiddenLocalTracksByPid=1299144.2-0&profileName=main&symbolServer=http%3A%2F%2F127.0.0.1%3A3000%2F28rfz7p9bj3skawjfq52bbwqyjaf96mlwny5pcm&thread=5&transforms=ff-67%7Emf-95%7Emf-97%7Emf-98%7Emf-35%7Emf-99%7Emf-100%7Emf-101%7Emf-102%7Emf-10%7Emf-11%7Emf-12%7Emf-87%7Emf-103%7Emf-91%7Emf-92%7Emf-93%7Emf-94&v=10), [PR](https://profiler.firefox.com/from-url/http%3A%2F%2F127.0.0.1%3A3001%2F0hqh983ljbln9kp2b2kmx5dbz81zc5lvbvg9n35%2Fprofile.json/calltree/?globalTrackOrder=0w9&hiddenGlobalTracks=0w35w9&hiddenLocalTracksByPid=1301351.2-0&profileName=none&symbolServer=http%3A%2F%2F127.0.0.1%3A3001%2F0hqh983ljbln9kp2b2kmx5dbz81zc5lvbvg9n35&thread=5&transforms=ff-99%7Emf-126%7Emf-128%7Emf-129%7Emf-136%7Emf-10%7Emf-11%7Emf-12%7Emf-127%7Emf-34%7Emf-125%7Emf-121%7Emf-132%7Emf-133%7Emf-134%7Emf-137%7Emf-135&v=10)). 

What I spotted is  that this branch now spends 5.6% of total time inside `infer_function_definition_statement` where we only spent 1.6% before. 

Expanding the tree shows that the difference comes from `infer_decorators` that isn't present in the profile for main. I suspect that it comes from 

```py
@lru_cache(maxsize=None)
def cached_tz(hour_str: str, minute_str: str, sign_str: str) -> timezone:
    sign = 1 if sign_str == "+" else -1
    return timezone(
        timedelta(
            hours=sign * int(hour_str),
            minutes=sign * int(minute_str),
        )
    )
```

Inside `infer_decorators` there's now a call to `Type::none` which starts semantic indexing and what not of `lru_cache. 

I tried to debug if we now infer a more accurate type for `lru_cache` but both main and this PR infer `Todo` for 

```py
from functools import lru_cache
from typing import reveal_type

reveal_type(lru_cache(maxsize=None))
```

The main thing that's puzzling me.... Why is the `lru_cache` type not cached? Shouldn't this bail out early instead of calling into `semantic_index`?


image


Ohh... never mind. I looked at the warmup phase instead of the incremental path. 

This is for the incremental part

image

The finding is the same... and, for some reason, `KnownClass::none` is re-executing semantic indexing of the `functools` module which it should not... Not sure why. We should look into this

**Comment by MichaReiser** [comment]
Looking at the salsa code. What this suggests is that we don't reuse the same salsa ids and, because of that have cache misses...

**Comment by MichaReiser** [comment]
I'm not sure what's happening and/or if this is a bug in the benchmark. I copied the tomllib and manually ran the red knot CLI in watch mode and the salsa events indicate that it uses the memoized values except for `_re.py` that changed.

**Comment by MichaReiser** [comment]
Okay, I might still have looked at the wrong profiles. I now created a PR to use a named closure for the incremental and warmup vs to avoid this in the future. 

* [incremental none](https://share.firefox.dev/48y8LsQ)
* [incremental main](https://share.firefox.dev/3YMAAum)

I haven't looked at it too closely because salsa's maybe changed after profiles are a bit of a pain but I strongly suspect that the difference mainly comes from that salsa now has to validate the ingredients for the module defining `None`?  I suspect that addressing https://github.com/astral-sh/ruff/issues/13169 could help here as well.

**Comment by carljm** [comment]
> the difference mainly comes from that salsa now has to validate the ingredients for the module defining `None`

Is this a speculation or is there evidence pointing to it? I couldn't really follow which comments here are based on looking at the wrong part of the profile :)

This could be, though. I was thinking we would already have a dependence on `types` module (because `builtins` imports it), but it's possible tomllib didn't actually resolve the type of any builtin depending on `types` module, so our lazy per-definition inference might have meant we previously didn't have to semantic-index `types` module.

If this is the cause of the regression, it's mostly just a weakness in our benchmark. Real-world codebases are likely to already rely on the `types` module in some way; adding a core typechecking dependency on it should not really be considered a real-world regression, if the regression just comes from "now we have to index that module."

I still think it's worth seeing whether Salsa-caching the resolution of the type of `None` makes a difference here. If the regression is entirely due to having all the new ingredients from the `types` module, then it won't.

**Comment by sharkdp** [comment]
## Performance investigation

- The original version of this PR had a -4% performance regression on the "incremental" *as well as the "cold" benchmark* ([results](https://codspeed.io/astral-sh/ruff/runs/6723f0d664ef40e4a982d256)). The "cold" benchmark was probably just slightly below the warning/error threshold.
- Consequently, a rebase to `main` (after the merge of https://github.com/astral-sh/ruff/pull/14034 by @MichaReiser) did not help with improving performance. Both versions are still at -4% ([results](https://codspeed.io/astral-sh/ruff/runs/6728abafa3e6c10bfecc5442))
- I then tried the suggestion by @carljm and added explicit salsa caching for the lookup of `NoneType` in 503aa4644d303ee8c5f411af6a07bd44b3108549. The CI run now passes, but the performance regression is still there with -3% for both versions ([results](https://codspeed.io/astral-sh/ruff/runs/6728b047a3e6c10bfecc544d)). Not sure if the resolution of these benchmarks is really sub-1%? If so, this is a small improvement.

**Review by MichaReiser — APPROVED** [review]

**Review by AlexWaygood — APPROVED** [review]
I've only skimmed, but this all LGTM, and I agree that a small perf "regression" here is fine given your analysis (and given the improvements to maintainability this gives us)

**Pull request merged**
